### DISCUSSION TITLE
Form submission

### DISCUSSION THREAD

**Discussion by leevigraham** [unpoly/unpoly#367] [Q&A] [answered]
Is there a high level overview of how unpoly submits forms? I doesn't look like it's using FormData to collect the inputs and send them … unless I'm missing something (probably).

I've also had some issues calling `.submit()` on the form directly where I expected unpoly to do it's `up-submit` functionality.

**Comment by triskweline** [comment]
> Is there a high level overview of how unpoly submits forms? I doesn't look like it's using FormData to collect the inputs and send them

See [`up.Params.fromForm()`](https://unpoly.com/up.Params.fromForm). It does not use `FormData`.

> I've also had some issues calling .submit() on the form directly where I expected unpoly to do it's up-submit functionality.

Use `up.submit(form)` for that.

**Comment by johndoe46** [comment]
Hello,

Is there any plan for a better support of webcomponents in forms for Unpoly ? Example: using Shoelace components in a form results in Unpoly being unable to fetch the data because the actual inputs are hidden in the shadow dom. For a standard form submit, Shoelace does the formData trick based on form events (see https://shoelace.style/getting-started/form-controls ), but since Unpoly does not use formData I'm not sure how I could use Unpoly declaratively with Shoelace (or any other UI lib based on shadow dom) components. Maybe I can use JS directly ?

**Reply by johndoe46** [reply]
Maybe I could up.watch elements and set a standard hidden input whenever changes occur ?

**Reply by triskweline** [reply, marked as answer]
Try this:

- Have your custom input element expose `{ name }` and `{ value }` properties (AFAIK Shoelance components already do)
- Make Unpoly aware of custom input elements by pushing a selector to [`up.form.config.fieldSelectors`](https://unpoly.com/up.form.config#config.fieldSelectors).

**Reply by johndoe46** [reply]
Works fine, excellent ! Quite a game changer for me :-)

btw, thanks for Unpoly, excellent library.
